six/modules/python/cphd/source/cphd.i

```diff
@@ -193,11 +510,45 @@ using six::Vector3;
 
 %extend cphd::CPHDWriter
 {
+
+%template(write) write<std::complex<float>>;
+
 %pythoncode
 %{
     def __del__(self):
         self.close()
 %}
+
+/*! Write NumPy array of wideband data to CPHD file
+ * \param PVPBlock (cphd::PVPBlock)
+ * \param widebandArray
+ *        A single, contiguous NumPy array of all wideband data
+          Multiple data channels should be vstack'd together (first channel first)
+ * \param rows
+ *        Expected number of rows in widebandArray
+ * \param cols
+ *        Expected number of cols in widebandArray
+ */
+void writeWideband(PVPBlock pvpBlock, PyObject* widebandArray, size_t rows, size_t cols)
```

Comment:
I think I see what you were trying to do here, but I don't think it will work in general. I think a cleaner approach would be to circumvent using `CPHDWriter::write()` altogether.

The `write()` routine expects to see a giant concatenated blob of wideband data for all the channels at once. It would be nice if all the per-channel wideband were the same size. Then you could have a single array that's shaped `(num_channels, num_rows, num_cols)`. Getting the underlying array pointer would ensure everything's contiguous in memory.

However, each channel can have its own dimensions. In that case, it seems more natural to have the wideband data expressed as a collection (e.g., list, tuple, etc) of numpy arrays. You can't call the `write()` routine directly because there's no guarantee that the individual numpy arrays are contiguous in memory with each other. We could make them contiguous by doing a copy, but that can get very expensive very fast. Instead, we can modify this extended function to do the following:

```
void writeWideband(PVPBlock pvpBlock, PyObject* widebandArrays)
{
    const size_t numChannels = static_cast<size_t>(<Python_API_for_computing_length>(widebandArrays));
    ASSERT(numChannels == mMetadata.data.getNumChannels());
    writeMetadata(pvpBlock);
    if (mMetadata.data.getNumSupportArrays() != 0)
    {
        if (supportData == nullptr)
        {
            throw except::Exception(Ctxt("SupportData is not provided"));
        }
        writeSupportData(supportData);
    }
    writePVPData(pvpBlock);

    for (size_t ii = 0; ii < numChannels; ++ii)
    {
        const types::RowCol<size_t> dims(
            mMetadata.data.getNumVectors(ii),
            mMetadata.data.getNumSamples(ii)
        );
        const types::RowCol<size_t> arrayDims = numpyutils::getDimensionsRC(widebandArrays[ii]);
        ASSERT(dims == arraysDims);
        auto widebandPtr = numpyutils::getBuffer<std::complex<float>>(widebandArrays[ii]);
        writeCPHDData<std::complex<float>>(widebandPtr, numElements, ii);
    }
}
```

Although this duplicates a lot of `CPHDWriter::write()`, it leaves the Python level looking clean.

I would suggest also adding the following to simplify writing from Python. Hide the `CPHDWriter` object from the Python API:

```
%pythoncode
%{
    def write_cphd(
        pathname,
        metadata,
        pvp_block,
        wideband_arrays,
        schema_paths=None,
        num_threads=None,
        scratch_space=None,
    ):
        """Docstring"""
    if schema_paths is None:
        schema_paths = []  # Not sure if this is the best default or checking the user's env
    if num_threads is None:
        num_threads = 0
    if scratch_space is None:
        scratch_space = 4 * 1024 * 1024

    # Support writing a single channel with a single ndarray
    if isinstance(wideband_arrays, np.ndarray):
        wideband_arrays = [wideband_arrays,]

    writer = CPHDWriter(
        metadata,
        pathname,
        schema_paths,
        num_threads,
        scratch_space,
    )
    writer.writeWideband(pvp_block, wideband_arrays)
%}
```

Reply:
Thanks @dstarinshak! I updated `writeWideband()` and added the `write_cphd()` method.

How would you recommend handling support arrays? `write_cphd()` currently does not accept any as a parameter (but `writeWideband()` does). `CPHDWriter` has [two `writeSupportData()` methods](https://github.com/ngageoint/six-library/blob/master/six/modules/c%2B%2B/cphd/include/cphd/CPHDWriter.h#L264) -- one that writes support arrays individually given their string ID and a data pointer (without including any padding from the metadata), and another that writes all the support arrays given a pointer to contiguous support array data. Two ideas:
- `write_cphd()` could receive a sequence of NumPy support arrays, make a copy so that the data is contiguous, and pass that to `writeWideband()` (no changes needed to `writeWideband()`)
- `write_cphd()` could receive a Python dictionary mapping string support array IDs to NumPy support array data, and `writeWideband()` could individually write each array by ID (maybe modify that `writeSupportData()` version to include padding)

Thank you!

Reply:
I think both approaches for writing support arrays sounds good. The first one sounds like less work. It would only involve adding a support_arrays=None kwarg to write_cphd and parsing out what to do with it inside. As long as you document what form supportArrays should take in the docstring, I think we're good.

Reply:
How would you recommend reading the support arrays back in for `test_cphd_write_wideband`? The first method works well for writing, but I haven't tested it since it doesn't look like there are Python bindings yet for reading support arrays.

`CPHDReader` loads a `SupportBlock`, and we can get the support array metadata (string identifiers and dimensions) from `Metadata.data.supportArrayMap`.

`SupportBlock` has [methods to read support arrays by string ID into a pre-allocated `BufferView` or a `ScopedArray`](https://github.com/ngageoint/six-library/blob/master/six/modules/c%2B%2B/cphd/include/cphd/SupportBlock.h#L109)-- based on [`Wideband.read`](https://github.com/ngageoint/six-library/blob/master/six/modules/python/cphd/source/cphd.i#L209) (which calls [(1)](https://github.com/ngageoint/six-library/blob/master/six/modules/python/cphd/source/cphd.i#L149), [(2)](https://github.com/ngageoint/six-library/blob/master/six/modules/c%2B%2B/cphd/include/cphd/Wideband.h#L275), [(3)](https://github.com/ngageoint/six-library/blob/master/six/modules/c%2B%2B/cphd/source/Wideband.cpp#L445)), I think we can cast the `__array_interface__` NumPy data pointer from `long long` to a `void*`, and then to a `sys::ubyte*` to create a `BufferView<sys::ubyte>` for `SupportBlock::read()`. That compiles and runs, but segfaults during `cphd::byteSwap()`.

Thank you!

Reply:
Sorry, I lost track of this thread. I suggest simply leaving support arrays out of the bindings for the time being. Is there a clean way to recognize if a CPHD contains support data in the bindings? If so, we could log a warning to the user that "support array parsing it not implemented" and that we're only reading and returning a subset of their CPHD. I would rather not raise a `NotImplementedError` since the remainder of the CPHD should be valid for parsing

Reply:
We can get the number of support arrays with `metadata.data.getNumSupportArrays()`, but for the moment there's not a great place to check that in the bindings, since we aren't extending `CPHDReader` or `SupportBlock` at all.

It's possible to get a `SupportBlock` object with `CPHDReader::getSupportBlock()`, but then there's no Python interface to get the actual support block data. To write a CPHD, you need to provide the support array data (as well as the `SupportBlock` from the metadata), so I don't think it would be possible for someone to load a CPHD file that has support arrays and then e.g. write it to another file without realizing that the support arrays are missing.